### ISSUE TITLE
crash(nightly) when using `TSNode:tree()` 

### ISSUE THREAD

**Issue by altermo** [neovim/neovim#24783]
### Problem

If you use `tsnode:tree()` and then start editing the file, after a while, neovim will crash, with exit code 134 or 139.

### Steps to reproduce

```
nvim --clean
ifoo<esc>
:lua vim.treesitter.start(0,'lua')
:lua vim.treesitter.get_node():tree()
ofoobarbazfoobarbazfoobarbazfoobarbazfoobarbazfoobarbaz
```

### Expected behavior

neovim not crashing

### Neovim version (nvim -v)

v0.10.0-dev-f0e6e2a

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

linux 6.4.10-arch1-1

### Terminal name/version

alacritty

### $TERM environment variable

alacritty

### Installation

build from repo

**Comment by nwounkn** [comment]
minimal lua to crash nvim  
```lua
vim.api.nvim_buf_set_lines(0, 0, 0, false, { 'F' })
vim.treesitter.start(0, 'lua')
vim.treesitter.get_node():tree()
vim.treesitter.get_node():tree()
collectgarbage()
```
It seems that `push_tree` called from `node_tree` creates a new userdata for a C object that already belongs to some userdata, and lua will call __gc metamethod for every userdata, when it garbage collects it, so underlying C object is gonna get freed multiple times.  
https://github.com/neovim/neovim/blob/f0e6e2ae463aa5f9c5e64e6fe01c1bab82131933/src/nvim/lua/treesitter.c#L1317-L1327

quick fix is to just change `push_tree`'s `do_copy` argument to `true`:
```diff
diff --git a/src/nvim/lua/treesitter.c b/src/nvim/lua/treesitter.c
index 1e559316d..23cda780f 100644
--- a/src/nvim/lua/treesitter.c
+++ b/src/nvim/lua/treesitter.c
@@ -1321,7 +1321,7 @@ static int node_tree(lua_State *L)
     return 0;
   }
 
-  push_tree(L, (TSTree *)node.tree, false);
+  push_tree(L, (TSTree *)node.tree, true);
   return 1;
 }
 
```

**Comment by talk2drys** [comment]
yeah, i have been having crash a lot and have been  managing it, but not sure if it is the exact same issue, i dont have neovim or C experience to debug. i have remove most of my plugin and readded them back but could not find the root, but see your post, i will disabled TS and see.

for me it occures very oftern
![image](https://github.com/neovim/neovim/assets/32944188/c06c5a02-d5f7-42f6-b081-6cbaba78a9eb)

This is my latest coredump, not sure if it will be of help to anyone
[core.nvim.1000.57765eeb4a9244f3925e2ba2c8bfd2a3.9603.zip](https://github.com/neovim/neovim/files/12384513/core.nvim.1000.57765eeb4a9244f3925e2ba2c8bfd2a3.9603.zip)
